server/src/main/java/org/opensearch/search/backpressure/SearchBackpressureService.java

```diff
@@ -213,7 +215,7 @@ TaskCancellation getTaskCancellation(CancellableTask task) {
         List<Runnable> callbacks = new ArrayList<>();
 
         for (TaskResourceUsageTracker tracker : taskResourceUsageTrackers) {
-            Optional<TaskCancellation.Reason> reason = tracker.cancellationReason(task);
+            Optional<TaskCancellation.Reason> reason = tracker.checkAndMaybeGetCancellationReason(task);
```

Comment:
Thanks for this, but in general we should think about decoupling tracking and action once thresholds have breached. Today it might be search cancellation but I do envision this as an action that modifies threadpool size/queue in a manner that creates a backpressure

Reply:
Also cancellation isn't truly back pressure :) it's load shedding

Reply:
That makes sense. Trackers can _recommend_ actions once thresholds are met, and cancellation of tasks can be one such action. This will however influence how dissimilar actions from different trackers are grouped/compared with each other in the `SearchBackpressureService`. For example, we need to aggregate the cancellation scores from each tracker before we start cancelling tasks. With generic actions, this might become really complicated.

Let's do a detailed design of this first and refactor as a follow-up. Enhancement: https://github.com/opensearch-project/OpenSearch/issues/4985